azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import abc
+import six
+from collections import deque
+from . import pipeline_ops
+from . import pipeline_events
+from . import mqtt_topic_helper
+from azure.iot.device.iothub.models import Message
+from azure.iot.device.common.transport.mqtt.mqtt_provider import MQTTProvider
+from azure.iot.device.common.transport import constant
+
+logger = logging.getLogger(__name__)
+
+# TODO: standardize on {} instead of %s
+# TODO: standardize logs with {}({}) prefix
+
+
+@six.add_metaclass(abc.ABCMeta)
+class PipelineStage(object):
+    """
+    Base class representing a stage in the processing pipeline.  Each stage is responsible for receiving
+    PipelineOperation objects from the top, possibly processing them, and possibly passing them down.  It
+    is also responsible for receiving PipelineEvent objects from the bottom, possibly processing them, and
+    possibly passing them up.  In both cases (PipelineOperation and PipelineEvent), the default behavior
+    should be to pass along an operation or event that you don't handle.
+    """
+
+    def __init__(self):
+        self.name = self.__class__.__name__
+        self.next = None
+        self.previous = None
+        self.pipeline_root = None
+
+    def run_op(self, op):
+        """
+        Run the given operation.  This is the public function that outside callers would call to run an
+        operation.  Derived classes should override the _run_op function to impliment stage-specific behavior.
+        """
+        try:
+            self._run_op(op)
+        except Exception as e:
+            logger.error(msg="Error in _run_op call", exc_info=e)
+            op.error = e
+            self.complete_op(op)
+
+    def run_ops_serial(self, *args, **kwargs):
+        """
+        Run the operations passed in *args in a serial manner, such that each operation waits for the
+        previous operation to be completed before running.  If an operation fails, the operations that
+        follow will not be run, but the operation passed using the finally_op will be run.  After
+        all operations and the finally_op (if provided) are run, then the function passed in the
+        callback parameter will be called, passing the last operation (in the event of success)
+        or the failing operation (in the event of failure).
+        """
+        worklist = deque(args)
+        finally_op = kwargs.get("finally_op", None)
```

Comment:
>finally_op = kwargs.get("finally_op", None) [](start = 8, length = 43)

Why not just let this be the last arg provided in *args? I'm not understanding the point of it as a keyword

Reply:
It's like try/except/finally.  a failure in the chain will skip the following operations, but the finally op will always be run.

---
In reply to: [279651953](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279651953) [](ancestors = 279651953)